hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +49,320 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+                                    access_location::host,
+                                    access_mode::readwrite);
+    ArrayHandle<Scalar4> h_angmom(m_pdata->getAngularMomentumArray(),
+                                  access_location::host,
+                                  access_mode::readwrite);
+    ArrayHandle<Scalar4> h_orientation(m_pdata->getOrientationArray(),
+                                       access_location::host,
+                                       access_mode::read);
+    ArrayHandle<Scalar3> h_inertia(m_pdata->getMomentsOfInertiaArray(),
+                                   access_location::host,
+                                   access_mode::read);
+    ArrayHandle<unsigned int> h_body(m_pdata->getBodies(),
+                                     access_location::host,
+                                     access_mode::read);
+    ArrayHandle<unsigned int> h_rtag(m_pdata->getRTags(), access_location::host, access_mode::read);
+    // add accumulated momentum to the central particle
+    unsigned int num_total = m_pdata->getN();
+    for (unsigned int idx = 0; idx < num_total; ++idx)
+        {
+        // check that the particle is in a rigid body and a central particle
+        const unsigned int central_tag = h_body.data[idx];
+        if (central_tag >= MIN_FLOPPY)
+            {
+            continue;
+            }
+        const unsigned int central_idx = h_rtag.data[central_tag];
+        if (central_idx != idx)
+            {
+            continue;
+            }
+
+        // get accumulated momentum for particle
+        const Scalar3 linmom_accum(h_linmom_accum.data[idx]);
+        const vec3<Scalar> angmom_accum(h_angmom_accum.data[idx]);
+
+        // compute and store new velocity
+        Scalar4 vel_mass = h_velocity.data[idx];
+        const Scalar mass = vel_mass.w;
+        if (mass > 0)
+            {
+            vel_mass.x += linmom_accum.x / mass;
+            vel_mass.y += linmom_accum.y / mass;
+            vel_mass.z += linmom_accum.z / mass;
+            h_velocity.data[idx] = vel_mass;
+            }
+
+        // compute new angular momentum
+        quat<Scalar> angmom(h_angmom.data[idx]);
+        const quat<Scalar> orientation(h_orientation.data[idx]);
+
+        // convert angular momentum to quaternion and update
+        const vec3<Scalar> inertia(h_inertia.data[central_idx]);
+        vec3<Scalar> angmom_change_body = rotate(conj(orientation), angmom_accum);
+        if (inertia.x == Scalar(0))
+            {
+            angmom_change_body.x = Scalar(0);
+            }
+
+        if (inertia.y == Scalar(0))
+            {
+            angmom_change_body.y = Scalar(0);
+            }
+
+        if (inertia.z == Scalar(0))
+            {
+            angmom_change_body.z = Scalar(0);
+            }
+        quat<Scalar> angmom_change = Scalar(2.0) * orientation * quat(0.0, angmom_change_body);
+        quat<Scalar> updated_angmom = angmom + angmom_change;
```

Comment:
Quaternions should have the `+=` operator implemented so you can do like this. Also, can you double check that this is the correct formula (including `orientation`) now that you've rotated the change in momentum to the body frame? It looks very similar to what you had before, so not sure if the old one or the new one is right.
```suggestion
        angmom += Scalar(2.0) * orientation * quat(0.0, angmom_change_body);
```

Reply:
I didn't actually have an equation for taking a space-frame angular momentum vector into a space-frame angular momentum quaternion. What I put before was a guess. The rotation to the body frame let's me instead use the steps I wrote out for going from a space-frame angular velocity to the angular momentum quaternion.

**Get a angular momentum quaternion from a desired space-frame angular velocity**
1. Rotate into the body frame ω^{b} = A(q)w^{s}
2. Multiply by moment of inertia tensor to get angular momentum J = Iω
3. Promote angular momentum to quaternion with 0 real part. J^{(4)} = (0, J)
4. Multiply by orientation q and by 2. Ps = 2q × J^{(4)}

Reply:
Very good, this seems correct to me then!